### PR TITLE
Replace ifdefs in net_context

### PR DESCRIPTION
**Pull request by pfl**
This patch set clears up net_context by replacing #ifdefs with IS_ENABLED(). This causes IPv6 and IPv4 functions to be seen by the compiler, so the function prototypes need to be defined so that the preprocessor can remove the functions that are now turned into simple return values by the preprocessor. The worse part is that both IPv4 and IPv6 addressing is now visible to the preprocessor and compiler, meaning that we need to come up with a solution in order not to reserve memory for IPv6 address sizes where only IPv4 is used.

This is one cleanup for #8725.

### CODE REVIEW

File: include/net/net_if.h [hunk 1]
```diff
@@ -1266,9 +1269,50 @@ void net_if_ipv6_dad_failed(struct net_if *iface, const struct in6_addr *addr);
  * @return Pointer to IPv6 address, NULL if not found.
  */
 struct in6_addr *net_if_ipv6_get_global_addr(struct net_if **iface);
+
 #else
-#define net_if_ipv6_select_src_addr(...)
+#define net_if_config_ipv6_get(...) -EOPNOTSUPP
```
Comment (jukkar) [inline include/net/net_if.h]: These defines do not seem to belong to this patch called "net_if: Define void net_if.h functions for CONFIG_NET_OFFLOAD"

File: include/net/net_ip.h [hunk 2]
```diff
@@ -162,12 +162,8 @@ struct sockaddr_storage {
 struct net_addr {
 	sa_family_t family;
 	union {
-#if defined(CONFIG_NET_IPV6)
```
Comment (jukkar) [inline include/net/net_ip.h]: I think this is bad idea because it uses more memory now. Even thou using #ifdefs looks ugly, there is a place for them here.

File: subsys/net/ip/net_context.c [hunk 3]
```diff
@@ -243,11 +241,10 @@ int net_context_get(sa_family_t family,
 				break;
 			}
 		}
-#endif
 
-#if defined(CONFIG_NET_CONTEXT_SYNC_RECV)
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: We should try to minimize storage in net_context struct, so here having #ifdef makes sense.

**Review by jukkar — CHANGES_REQUESTED** [review]
Most of the stuff looks ok, but I am against those changes that would increase the size of various structs like net_context.

File: include/net/net_ip.h (hunk 2 above)
Comment (pfalcon) [reply include/net/net_ip.h]: +1, Except there's nothing "ugly" in #ifdef's, they're beautiful. There's a way to make them even more beautiful though - indent them, just like normal code is being indented. (Well, some conventions should be set, e.g. here "#if" could start either at "union" indent or "struct in6_addr in6_addr" indent).

Comment (nordic-krch) [reply include/net/net_ip.h]: #ifdefs looking ugly is not the worst thing. Dead code is. With ifdefs you can have a lot of code which is not touched by the compiler until someone, someday enables the feature and finds out that it does not compile.

Personally, i tried to avoid #ifdefs as much as possible and there are ways to keep memory usage low.

In that case it's all about the size of struct in6_addr so it could be reduced to a pointer:
```
union {
#if defined(CONFIG_NET_IPV6)
		u8_t u6_addr8[16];
		u16_t u6_addr16[8]; /* In big endian */
		u32_t u6_addr32[4]; /* In big endian */
#else 
		u8_t *u6_addr8;
		u16_t *u6_addr16; /* In big endian */
		u32_t *u6_addr32; /* In big endian */
#endif 
	} in6_u;
```

File: include/net/net_if.h (hunk 1 above)
Comment (pfl) [reply include/net/net_if.h]: These function pop out since the preprocessor does not anymore remove the code, but instead the compiler sees it now with IS_ENABLED() being used.

File: subsys/net/ip/net_context.c (hunk 3 above)
Comment (pfl) [reply subsys/net/ip/net_context.c]: Except that other code now previously removed by #ifdefs in the preprocessor will point towards in6 and in members in this struct. So we need to define the in6 member so that the compiler is happy and no extra memory is used.

Comment (pfl) [reply subsys/net/ip/net_context.c]: > Isn't there compile error now as in net_context struct the recv_data_wait semaphore is surrounded by #ifdef's, and this PR does not touch net_context.h?

At least I didn't see one with 'sanitycheck -t net', but even if it survives the compiler and test case(s) there still might be an issue lurking.

File: include/net/net_ip.h (hunk 2 above)
Comment (pfl) [reply include/net/net_ip.h]: Let's try with that.

Comment (pfl) [reply include/net/net_ip.h]: > Except there's nothing "ugly" in #ifdef's, they're beautiful.

YMMV, but #ifdefs hide away code paths (as nordic-krch pointed out) and are therefore an unnecessary evil to wrestle with. And the issue two levels up explicitly asked for #ifdef removal.

**Comment by pfl** [comment]
Updated with the pointer solution proposed by nordic-krch. Now the address structures won't grow as a result of adding IPv6 addressing.

File: include/net/net_ip.h (hunk 2 above)
Comment (pfalcon) [reply include/net/net_ip.h]: @nordic-krch 

> Dead code is. With ifdefs you can have a lot of code which is not touched by the compiler until someone, someday enables the feature and finds out that it does not compile.

That's very obvious problem with very obvious solution - have a good CI with good coverage. Zephyr exactly has that. Otherwise, saying that is about the same as saying "if's are bad - suppose some branch is not executed, until someone, someday. And then [fighters flip upside down on crossing equator](https://forums.theregister.co.uk/forum/1/2007/02/28/f22s_working_again/)". Solution is not killing *all* if's, but again, have good testing coverage.

> Personally, i tried to avoid #ifdefs as much as possible and there are ways to keep memory usage low.

Of course, everyone uses #ifdef's only when required, and doesn't use when not required.



Now let's look at the risks of using code like `if (1)`/`if (0)` (to which IS_ENABLED translates):

* It's effectively dead code, some compilers love to throw warnings at such, which can be turned to errors easily.
* Some overzealous static analysis tools can do the same. Coverity which we use seems to be smart not to (or was ~~patched~~ configured not to), but who knows what will be used next.
* Finally, there're creative coding style standards. One of the most creative known is MISRA C. With all the mad things it required, it would be a wonder if it allowed dead if's.


So, with no-zero probability, all this fashion-motion to go from #ifdef to if(IS_ENABLED()) can go into reverse direction.

**Review by pfalcon — CHANGES_REQUESTED** [review]
Copied from here: https://github.com/zephyrproject-rtos/zephyr/pull/9752#discussion_r215080169 , by @pfl :

> And the issue two levels up explicitly asked for #ifdef removal.

Ok, let's see. You apparently refer to https://github.com/zephyrproject-rtos/zephyr/issues/8723, which has a phrase:

>> "Reduce #ifdefs usage inside functions"

In which language, in whose interpretation "reduce" means "exterminate ALL"?

>  #ifdefs ... are therefore an unnecessary evil to wrestle with.

It seems the situation is the opposite - you assigned some imaginable negative traits to #ifdef's and trying to wrestle with exterminating them completely. It's impossible. There's no need.

Strong -1.

Suggestion to going forward with this:

1. Obviously, #ifdef's can be replaced with if() only in the *code*.
2. Based on allegedly faulty suggestions to do something to headers, as expressed in comments in this PR, it's probably worth to have an explicit suggestion of treating headers as invariants. If there's something wrong with headers, or idea of improvement is seen, let that be a separate issue (and preferrably an RFC ticket, and not a big red-green PR).

**Comment by jukkar** [comment]
I think we should try to leave the #ifdef's in structs in place. The proposal by @nordic-krch just looks too kludgy. Of course this means that we would have then #ifdef's in the code too. In order to make sure that the code in those #ifdef's is ok, we would also need to have suitable unit test so that those code paths get tested.

**Comment by pfl** [comment]
Ok, let's revisit this patch with that in mind.
